src/main/java/org/jetbrains/research/refactorinsight/ui/DiffHintLineMarkerProvider.java

```diff
@@ -0,0 +1,222 @@
+        }
+        return null;
+    }
+
+    private boolean fromSameFile(PsiElement element, RefactoringInfo refactoringInfo) {
+        String elementPath = element.getContainingFile().getVirtualFile().getPath();
+        String refactoringClassPath = refactoringInfo.getRightPath();
+        return elementPath.endsWith(refactoringClassPath);
+    }
+
+    private boolean isIdentifier(PsiElement element) {
+        Language elementLanguage = element.getLanguage();
+        if (elementLanguage.equals(JavaLanguage.INSTANCE)) {
+            return element instanceof PsiIdentifier;
+        } else if ("kotlin".equalsIgnoreCase(elementLanguage.getID())) {
+            return element instanceof LeafElement leaf && "IDENTIFIER".equals(leaf.getElementType().toString());
+        }
+        return false;
+    }
+
+    private static class RefactoringInfoHint extends MergeableLineMarkerInfo<PsiElement> {
+
+        RefactoringInfoHint(@NotNull final PsiElement element, Function<? super PsiElement, String> tooltipProvider, RefactoringInfo refactoringInfo) {
+            super(element,
+                    element.getTextRange(),
+                    RefactorInsightIcons.toggle,
+                    tooltipProvider,
+                    new MyIconGutterHandler(refactoringInfo),
+                    GutterIconRenderer.Alignment.LEFT,
+                    () -> tooltipProvider.fun(element));
+        }
+
+        @Override
+        public boolean canMergeWith(@NotNull MergeableLineMarkerInfo<?> info) {
+            return info instanceof RefactoringInfoHint;
+        }
+
+        @Override
+        public Icon getCommonIcon(@NotNull List<? extends MergeableLineMarkerInfo<?>> infos) {
+            return RefactorInsightIcons.toggle;
+        }
+
+        @NotNull
+        @Override
+        public Function<? super PsiElement, String> getCommonTooltip(@NotNull List<? extends MergeableLineMarkerInfo<?>> infos) {
+            return __ -> "Refactoring detected";
+        }
+    }
+
+    private static class MyIconGutterHandler implements GutterIconNavigationHandler<PsiElement> {
+
+        private final RefactoringInfo refactoringInfo;
+
+        MyIconGutterHandler(RefactoringInfo refactoringInfo) {
+            this.refactoringInfo = refactoringInfo;
+        }
+
+        @Override
+        public void navigate(MouseEvent e, PsiElement nameIdentifier) {
+            final PsiElement listOwner = nameIdentifier.getParent();
+            final PsiFile containingFile = listOwner.getContainingFile();
+            final VirtualFile virtualFile = PsiUtilCore.getVirtualFile(listOwner);
+
+            if (virtualFile != null && containingFile != null) {
+                final JBPopup popup = createTextPopup(nameIdentifier.getProject());
+                popup.show(new RelativePoint(e));
+            }
+        }
+
+        @NotNull
+        private JBPopup createTextPopup(Project project) {
+            return JBPopupFactory.getInstance()
```

Comment:
Popup with refactoring description and diff button **doesn't disappear** after clicking on the button. So, user needs to click somewhere to make it disappear. Maybe we should create something else instead of JBPopup? Please, investigate what could be used in `JBPopupFactory`.

Reply:
In the implementation with list refactorings, when clicking on a list element, a diff opens and the list disappears - https://github.com/JetBrains-Research/RefactorInsight/issues/97

Reply:
Great!👍